<a href="https://colab.research.google.com/github/juanpalominoh/Fundamentos-de-Econometria/blob/main/Tema_6_Estimaci%C3%B3n_sujeta_a_Restricciones_Lineales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Estimación sujeta a Restricciones Lineales

Tenemos información de producción, capital y trabajo de 27 empresas y deseamos estimar la función de producción. Hay dos modelos alternaitivos para estimar la función de producción Cobb-Douglas logaritmica:
\begin{gather*}
ln(Q_{i}) = \beta_{1} + \beta_{2} ln(K_{i}) + \beta_{3} ln(L_{i}) + \epsilon_{i}
\end{gather*}
Y la translogarítmica:
\begin{gather*}
Q_{i} = \alpha_{1} + \alpha_{2} ln(K_{i}) + \alpha_{3} ln(L_{i}) + \alpha_{4} [\frac{1}{2}ln(K_{i})]^2 + \alpha_{5} [\frac{1}{2}ln(L_{i})]^2 + \alpha_{6} ln(K_{i})*ln(L_{i}) + \epsilon_{i}
\end{gather*}

In [ ]:
install.packages("AER")
library(AER)

In [ ]:
data(SIC33)
SIC33

output  labor   capital 
1   657.29  162.31   279.99
2   935.93  214.43   542.50
3  1110.65  186.44   721.51
4  1200.89  245.83  1167.68
5  1052.68  211.40   811.77
6  3406.02  690.61  4558.02
7  2427.89  452.79  3069.91
8  4257.46  714.20  5585.01
9  1625.19  320.54  1618.75
10 1272.05  253.17  1562.08
11 1004.45  236.44   662.04
12  598.87  140.73   875.37
13  853.10  145.04  1696.98
14 1165.63  240.27  1078.79
15 1917.55  536.73  2109.34
16 9849.17 1564.83 13989.55
17 1088.27  214.62   884.24
18 8095.63 1083.10  9119.70
19 3175.39  521.74  5686.99
20 1653.38  304.85  1701.06
21 5159.31  835.69  5206.36
22 3378.40  284.00  3288.72
23  592.85  150.77   357.32
24 1601.98  259.91  2031.93
25 2065.85  497.60  2492.98
26 2293.87  275.20  1711.74
27  745.67  137.00   768.59

Estimar la función logaritmica:

In [ ]:
fm_cb <- lm(output ~ capital + labor, data=log(SIC33))
summary(fm_cb)


Call:
lm(formula = output ~ capital + labor, data = log(SIC33))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.30385 -0.10119 -0.01819  0.05582  0.50559 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.17064    0.32678   3.582  0.00150 ** 
capital      0.37571    0.08535   4.402  0.00019 ***
labor        0.60300    0.12595   4.787 7.13e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.1884 on 24 degrees of freedom
Multiple R-squared:  0.9435,	Adjusted R-squared:  0.9388 
F-statistic: 200.2 on 2 and 24 DF,  p-value: 1.067e-15


In [ ]:
# Suma de los residuos al cuadrado
deviance(fm_cb)

[1] 0.8516337

In [ ]:
# Matriz de varianza-covarianza
vcov(fm_cb)

(Intercept) capital      labor       
(Intercept)  0.10678650  0.001188850 -0.019835398
capital      0.00118885  0.007283931 -0.009616201
labor       -0.01983540 -0.009616201  0.015864400

In [ ]:
# Usando matrices
X <- cbind(1,
          log(SIC33$capital),
          log(SIC33$labor))
y <- log(SIC33$output)

Tenemos la información de la siguiente matriz:
\begin{gather*}
(X'X)^{-1}
\end{gather*}

In [ ]:
# Inversa (X'X)^-1
invx <- solve(crossprod(X))
invx

[,1]        [,2]        [,3]      
[1,]  3.00936430  0.03350314 -0.5589840
[2,]  0.03350314  0.20526941 -0.2709954
[3,] -0.55898396 -0.27099541  0.4470767

## Pregunta 1

**Probar la hipótesis de rendimientos constantes a escala: $H_0: \beta_{2} + \beta_{3}=1$**

Definamos a 
\begin{gather*}
R= 
\begin{bmatrix}
0 & 1 & 1
\end{bmatrix}
\end{gather*}
y
\begin{gather*}
r=
\begin{bmatrix}
1
\end{bmatrix}
\end{gather*}

In [ ]:
# El vector de coeficientes es:
b_hat <- solve(crossprod(X)) %*% crossprod(X, y)
b_hat

[,1]     
[1,] 1.1706440
[2,] 0.3757103
[3,] 0.6029992

In [ ]:
# La matriz R
R <- as.matrix(cbind(0, 1, 1))
R

[,1] [,2] [,3]
[1,] 0    1    1

Entonces
\begin{gather*}
R\hat{\beta} - r= 
\begin{bmatrix}
0 & 1 & 1
\end{bmatrix}
\begin{bmatrix}
1.170644 \\
.03757101 \\
.6029995
\end{bmatrix}
-1=-0.021291
\end{gather*}

In [ ]:
# El cálculo es:
Rbr <- R%*%b_hat - 1
Rbr

[,1]       
[1,] -0.02129049

Luego
\begin{gather*}
R(X'X)^{-1}R= 
\begin{bmatrix}
0 & 1 & 1
\end{bmatrix}
\begin{bmatrix}
3.009364 & 0.033501 & -0.558981 \\
0.033501 & 0.205268 & -0.27099 \\
-0.558981 & -0.27099 & 0.447074
\end{bmatrix}
\begin{bmatrix}
0 \\
1 \\
1
\end{bmatrix}
= 0.110355
\end{gather*}

In [ ]:
# El cálculo es:
RXR <- (R %*% invx) %*% t(R)
RXR

[,1]     
[1,] 0.1103553

Para el cálculo del F estadístico:
\begin{gather*}
F = \frac{\frac{(R\hat{\beta}-r)^{'}[R(X'X)^{-1}R']^{-1}(R\hat{\beta}-r)}{q}}{\frac{\hat{\epsilon}^{'}\hat{\epsilon}}{n-K}}
\end{gather*}

In [ ]:
num <- (Rbr * (RXR^-1) * Rbr)/1
den <- deviance(fm_cb)/24
f <- num/den
f

[,1]     
[1,] 0.1157542

In [ ]:
# Valor crítico de la tabla F-Fisher
qf(0.95, 1, 24)

[1] 4.259677

Dado que el valor crítico de tabla F(1,24)=4.25 es mayor a F=0.116, no se rechaza la hipótesis nula de rendimientos constantes a escala.

In [ ]:
# Hipótesis de retornos constantes a escala
linearHypothesis(fm_cb, "labor + capital = 1")

Res.Df RSS       Df Sum of Sq   F         Pr(>F)   
1 25     0.8557412 NA          NA        NA        NA
2 24     0.8516337  1 0.004107506 0.1157542 0.7366458

## Pregunta 2

**Estimar el modelo Cobb-Douglas imponiendo la restricción de rendimientos constantes a escala**

Calculamos $\tilde{\beta}$ de acuerdo a:
\begin{gather*}
\tilde{\beta} = \hat{\beta} + (X'X)^{-1}R'(R(X'X)^{-1}R')^{-1}(r-R\hat{\beta})
\end{gather*}

In [ ]:
beta_rest <- b_hat + (invx %*% t(R)) %*% RXR^(-1) %*% abs(Rbr)
beta_rest

[,1]    
[1,] 1.069265
[2,] 0.363030
[3,] 0.636970

## Pregunta 3

Estimar lo mismo transformando el modelo original en uno restricto
\begin{gather*}
ln(Q_{i}) = \tilde{\beta}_{1} + \tilde{\beta}_{2} ln(K_{i}) + (1-\tilde{\beta}_{2}) ln(L_{i}) + \tilde{\epsilon}_{i}
\end{gather*}

\begin{gather*}
ln(Q_{i}) - ln(L_{i})= \tilde{\beta}_{1} + \tilde{\beta}_{2} [ln(K_{i}) - ln(L_{i})] + \tilde{\epsilon}_{i}
\end{gather*}

In [ ]:
# Se crea las variables producto per capita y capital per capita
SIC33$lnql <- log(SIC33$output) - log(SIC33$labor)
SIC33$lnkl <- log(SIC33$capital) - log(SIC33$labor)

In [ ]:
# Modelo restringido estimado
model_rest <- summary(lm(lnql ~ lnkl, data=SIC33))
model_rest


Call:
lm(formula = lnql ~ lnkl, data = SIC33)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.29281 -0.10015 -0.01333  0.05748  0.51776 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  1.06926    0.13176   8.115 1.81e-08 ***
lnkl         0.36303    0.07541   4.814 6.03e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.185 on 25 degrees of freedom
Multiple R-squared:  0.4811,	Adjusted R-squared:  0.4603 
F-statistic: 23.18 on 1 and 25 DF,  p-value: 6.03e-05


In [ ]:
cat("El valor de beta es", 1-model_rest$coefficients[2])

El valor de beta es 0.63697

## Pregunta 4
**Estimar los modelos Cobb Douglas y translogarítmico y decida qué modelo es más apropiado para los datos**

El modelo translogarítmico:
\begin{gather*}
Q_{i} = \alpha_{1} + \alpha_{2} ln(K_{i}) + \alpha_{3} ln(L_{i}) + \alpha_{4} [\frac{1}{2}ln(K_{i})]^2 + \alpha_{5} [\frac{1}{2}ln(L_{i})]^2 + \alpha_{6} ln(K_{i})*ln(L_{i}) + \epsilon_{i}
\end{gather*}

In [ ]:
## Modelo Translogarítmico
fm_tl <- lm(output ~ labor + capital + I(0.5*labor^2) + I(0.5*capital^2) + I(labor * capital),
          data = log(SIC33))
summary(fm_tl)


Call:
lm(formula = output ~ labor + capital + I(0.5 * labor^2) + I(0.5 * 
    capital^2) + I(labor * capital), data = log(SIC33))

Residuals:
     Min       1Q   Median       3Q      Max 
-0.33990 -0.10106 -0.01238  0.04605  0.39281 

Coefficients:
                   Estimate Std. Error t value Pr(>|t|)  
(Intercept)         0.94420    2.91075   0.324   0.7489  
labor               3.61364    1.54807   2.334   0.0296 *
capital            -1.89311    1.01626  -1.863   0.0765 .
I(0.5 * labor^2)   -0.96405    0.70738  -1.363   0.1874  
I(0.5 * capital^2)  0.08529    0.29261   0.291   0.7735  
I(labor * capital)  0.31239    0.43893   0.712   0.4845  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.1799 on 21 degrees of freedom
Multiple R-squared:  0.9549,	Adjusted R-squared:  0.9441 
F-statistic: 88.85 on 5 and 21 DF,  p-value: 2.121e-13


El modelo Cobb-Douglas es un caso especial del translogarítmico que impone la restricción $\alpha_{4}=\alpha_{5}=\alpha_{6}=0$. 

Probar la hipótesis $H_{0}: \alpha_{4}=\alpha_{5}=\alpha_{6}=0$.

Entonces, tomando el modelo Cobb Douglas como el modelo restringido, se calcula el estadístico F.

In [ ]:
fm_tl$coefficients

(Intercept)              labor            capital   I(0.5 * labor^2) 
        0.94419685         3.61363894        -1.89311291        -0.96405221 
I(0.5 * capital^2) I(labor * capital) 
        0.08529466         0.31238704

In [ ]:
linearHypothesis(fm_tl, c("I(0.5 * labor^2)=0", "I(0.5 * capital^2)=0", "I(labor * capital)=0"))

Res.Df RSS       Df Sum of Sq F        Pr(>F)   
1 24     0.8516337 NA        NA       NA        NA
2 21     0.6799272  3 0.1717064 1.767755 0.1841388

No se rechaza la hipótesis nula.